[#174860845] Add bonus-redeem

pasqualedevita opened the pull request:
Script to reprocess redeemed requests.

How to use (config **prod_bonus**, select **option 4**)
```./bin/run bonuses-bv:redeem-bonuses --day=DAY --fromTime=FROM_TIME --toTime=TO_TIME --apiKey=API_KEY```

Test command (config **dev**, select **option 2**)
```./bin/run bonuses-bv:redeem-bonuses --day=2020-09-23 --apiKey=API_KEY --apiUrl=https://httpbin.org/anything```

gunzip commented:
it's not clear if this command does both thing (download + reprocess) anyway it's called (the description inside is "Download redeemed bonuses"). can we 1) put a better description 2) eventually split this in two different commands if download-only is really needed?

gunzip commented on src/commands/bonuses-bv/redeem-bonuses.ts:
```diff
@@ -0,0 +1,133 @@
+          file: redeemedRequest,
+          statusCode: rawDataResponse.status.toString()
+        });
+        // pause 5 seconds
```
why? (can you add a comment?)

pasqualedevita replied:
changed to 1 second
added comment: pause 1 second to avoid 429 statusCode response

gunzip commented on src/commands/bonuses-bv/redeem-bonuses.ts:
```diff
@@ -0,0 +1,133 @@
+        },
+        {
+          printLine: this.log,
+          ...parsedFlags // parsed flags
```
```suggestion
          ...parsedFlags
```

pasqualedevita replied:
fixed, removed comment

gunzip commented on src/utils/azure.ts:
```diff
@@ -26,35 +30,77 @@ export const agid: IAzureConfig = {
-  configName: "io-dev-aks-k8s-01",
-  cosmosDatabaseName: "io-dev-sqldb-db-01",
+  configName: "io-d",
+  cosmosDatabaseName: "raimondo",
```
what? :)

pasqualedevita replied:
We finished free cosmos containers :) Anyway I fixed it with standard `db` container.
I think that `configName` is never used.

gunzip commented on src/utils/azure.ts:
```diff
@@ -26,35 +30,77 @@ export const agid: IAzureConfig = {
   storageMessagesContainer: "message-content",
-  storageName: "iodevsaappdata"
+  storageBonusRedeemedContainer: "redeemed-test",
+  storageName: "devpasqualesa"
```
?

pasqualedevita replied:
fixed and created `iodsttest` in dev environment

pasqualedevita commented:
> it's not clear if this command does both things (download + reprocess) anyway it's called (the description inside is "Download redeemed bonuses"). can we 1) put a better description 2) eventually split this in two different commands if download-only is really needed?

Thank you, I've fixed the description.
The first implementation is designed with two commands (download and reprocess) but I also think is better to have one command.

francescopersico commented on src/commands/bonuses-bv/redeem-bonuses.ts:
```diff
@@ -0,0 +1,136 @@
+          redeemedRequest,
+          "utf8"
+        ) as string;
+        const rawDataResponse = await fetch(parsedFlags.apiUrl, {
```
You are going to flood the server with a large amount of requests that contains a lot of redeemed bonus.
I think it can be useful to have a little delay between each request.

francescopersico replied:
Just noticed there is already a 1sec delay.

francescopersico replied:
Just noticed that there is already a 1sec delay.
I think it is safer to set a 5/10sec delay.

pasqualedevita replied:
setted to 10s

francescopersico commented on src/utils/azure.ts:
```diff
@@ -11,9 +11,13 @@ export interface IAzureConfig {
   cosmosNotificationContainer: string;
   cosmosNotificationStatusContainer: string;
   cosmosServicesContainer: string;
+  cosmosSenderServicesContainer: string;
+  cosmosUserDataProcessingContainer: string;
```
Are these related to this PR?

pasqualedevita replied:
removed

pasqualedevita commented:
1. reprocess redeemed requests without save requests as local file
1. setted pause to 10s
1. removed required attribute to flags with default value
1. renamed flag containerFolder in day
1. fixed descriptions and comments

pasqualedevita commented on README.md:
```diff
@@ -199,6 +200,25 @@ EXAMPLE
+  $ io-ops bonuses-bv:redeem-bonuses
+
+OPTIONS
+  -c, --containerFolder=containerFolder  (required) container folder
```
renamed in day

pasqualedevita commented on README.md:
```diff
@@ -199,6 +200,25 @@ EXAMPLE
+
+OPTIONS
+  -c, --containerFolder=containerFolder  (required) container folder
+  -k, --apiKey=apiKey                    (required) api key
```
fixed description

pasqualedevita commented on README.md:
```diff
@@ -199,6 +200,25 @@ EXAMPLE
+OPTIONS
+  -c, --containerFolder=containerFolder  (required) container folder
+  -k, --apiKey=apiKey                    (required) api key
+  -o, --tmpDir=tmpDir                    (required) [default: tmp] temp directory
```
removed required attribute when it's setted a default value.

pasqualedevita replied:
We can see the file without go on Azure and search it, but it is not a big benefit.
In the new version, the script reprocess the request on the fly without save it in temp directory.
Thank you :)

pasqualedevita commented:
added flags fromTime and toTime to filter redeemed requests

Undermaken commented on README.md:
```diff
@@ -199,6 +200,28 @@ EXAMPLE
+  --apiKey=apiKey      (required) bonus vacanze api key
+  --apiUrl=apiUrl      [default: https://api.io.italia.it/api/bonus-vacanze/v1/redeemed] bonus vacanze api url
+
+  --day=day            (required) filter reddem request from specified day (utc timezome, required format yyyy-MM-dd, ie
```
```suggestion
  --day=day            (required) filter redeemed request from specified day (utc timezone, required format yyyy-MM-dd, ie
```

pasqualedevita replied:
fixed

Undermaken commented on src/commands/bonuses-bv/redeem-bonuses.ts:
```diff
@@ -0,0 +1,202 @@
+  public static flags = {
+    day: flags.string({
+      description:
+        "filter reddem request from specified day (utc timezome, required format yyyy-MM-dd, ie 2020-05-25)",
```
```suggestion
        "filter redeemed request from specified day (utc timezone, required format yyyy-MM-dd, ie 2020-05-25)",
```

pasqualedevita replied:
fixed

Undermaken commented on src/commands/bonuses-bv/redeem-bonuses.ts:
```diff
@@ -0,0 +1,202 @@
+    const { flags: parsedFlags } = this.parse(RedeemBonuses);
+
+    const day = DateTime.fromFormat(
+      parsedFlags.day + " " + "utc",
```
```suggestion
      parsedFlags.day + " utc",
```

pasqualedevita replied:
fixed

Undermaken commented on src/commands/bonuses-bv/redeem-bonuses.ts:
```diff
@@ -0,0 +1,202 @@
+    let fromTime = undefined;
+    if (parsedFlags.fromTime !== undefined) {
+      fromTime = DateTime.fromFormat(
+        parsedFlags.day + " " + parsedFlags.fromTime + " " + "utc",
```
```suggestion
        `${parsedFlags.day} ${parsedFlags.fromTime} utc`,
```
[it's the same](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Template_literals) but it's more readable

pasqualedevita replied:
fixed, thank you!

pasqualedevita commented:
Reprocess redeemed bonuses finished. If it's ok we can merge into master
